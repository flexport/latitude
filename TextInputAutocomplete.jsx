/**
 * TEAM: frontend_infra
 * @flow strict
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import useDropdown from "./tools/useDropdown";
import TextInput from "./TextInput";
import DropdownList from "./select/DropdownList";
import PopupWithClickAway from "./popup/PopupWithClickAway";
import {zIndices} from "./tools/zIndices";

type Option = string;

type TextInputProps = React.ElementConfig<typeof TextInput>;

type Props = {|
  /** all TextInput props will be passed into the TextInput */
  ...TextInputProps,
  +suggestions: $ReadOnlyArray<Option>,
  /**
   * positively filters an suggestion according to the callback. Specify `null` to
   * not filter suggestions
   */
  +suggestionsFilter: ((query: string, option: Option) => boolean) | null,
  /** the maximum number of suggestions that will be presented */
  +maximumOptions: number,
  /** the suggestion to highlight for keyboard selection, can be a function that
   * returns what suggestion to highlight on value change */
  +highlight?: string | ((?string) => ?string),
|};

export const defaultFilter = (query: string, option: Option) => {
  const trimmedOption = option.toLowerCase().trim();
  const trimmedQuery = query.toLowerCase().trim();

  return trimmedOption.includes(trimmedQuery);
};

const defaultProps = {
  suggestionsFilter: defaultFilter,
  maximumOptions: 10,
};

/**
 * @category Data Entry
 * @short Collect simple text input from the user with dropdown suggestions
 * @brandStatus V2
 * @status Stable
 * TextInputAutocomplete is a text input with a suggestions dropdown. By default a maximum
 * of 10 suggestions are displayed by the dropdown.
 */
export default function TextInputAutocomplete({
  suggestions,
  suggestionsFilter,
  maximumOptions,
  value,
  highlight,
  ...textInputProps
}: Props) {
  const filteredSuggestions = suggestions
    .filter((option: string) =>
      suggestionsFilter ? suggestionsFilter(value, option) : true
    )
    .slice(0, maximumOptions);

  const popper = React.useRef();

  const handleChange = (newValue: string) => {
    textInputProps.onChange(newValue);

    if (typeof highlight === "function") {
      const valueToHighlight = highlight(newValue);

      if (valueToHighlight != null) {
        setHighlightedLabel(valueToHighlight);
      }
    }
  };

  const {highlightedIndex, handlers, setHighlightedLabel} = useDropdown(
    filteredSuggestions,
    {
      popper,
      onChange: handleChange,
      rememberHighlightPosition: suggestionsFilter === null,
      onMouseDown: e => {
        if (textInputProps.onClick) {
          textInputProps.onClick(e);
        }
      },
      onFocus: e => {
        if (textInputProps.onFocus) {
          textInputProps.onFocus(e);
        }
      },
      onBlur: e => {
        if (textInputProps.onBlur) {
          textInputProps.onBlur(e);
        }
      },
      onKeyDown: e => {
        if (textInputProps.onKeyDown) {
          textInputProps.onKeyDown(e);
        }
      },
    }
  );

  if (typeof highlight === "string") setHighlightedLabel(highlight);

  return (
    <PopupWithClickAway ref={popper}>
      {(Target, Popup) => (
        <div className={css(styles.container)}>
          <Target>
            <TextInput
              {...textInputProps}
              value={value}
              onChange={handlers.handleChange}
              onMouseDown={handlers.handleMouseDown}
              onFocus={handlers.handleFocus}
              onBlur={handlers.handleBlur}
              onKeyDown={handlers.handleKeyDown}
            />
          </Target>
          <Popup
            placement="bottom-start"
            zIndex={zIndices.zIndex1500AboveModal.value}
            noPortal={true}
          >
            <div
              tabIndex={-1}
              onFocus={handlers.handleFocus}
              onBlur={handlers.handleBlur}
              className={css(styles.dropdownContainer)}
            >
              <DropdownList
                options={filteredSuggestions.map(suggestion => ({
                  label: suggestion,
                }))}
                highlightedOption={
                  highlightedIndex != null
                    ? filteredSuggestions[highlightedIndex]
                    : null
                }
                onClick={handlers.handleItemClick}
              />
            </div>
          </Popup>
        </div>
      )}
    </PopupWithClickAway>
  );
}

TextInputAutocomplete.defaultProps = defaultProps;

const styles = StyleSheet.create({
  container: {
    position: "relative",
  },
  dropdownContainer: {
    position: "absolute",
    left: "0",
    top: "100%",
    padding: "6px 0",
    minWidth: "100%",
    zIndex: "10",
    transform: "none",
  },
});
