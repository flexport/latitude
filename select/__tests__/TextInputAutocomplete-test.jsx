/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {mount} from "enzyme";
import TextInputAutocomplete, {
  defaultFilter,
} from "../../TextInputAutocomplete";

function mountTextInputAutocomplete(props?: {}) {
  return mount(
    <TextInputAutocomplete
      value=""
      onChange={() => {}}
      suggestions={["aaa", "aaabbb", "aaabbbccc"]}
      // $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0)
      {...props}
    />
  );
}

describe("TextInputAutocomplete", () => {
  describe("default filter", () => {
    it("positively filters a query that exactly matches an option", () => {
      expect(defaultFilter("aaa", "aaa")).toBe(true);
    });

    it("positively filters a query that matches an option after trim", () => {
      expect(defaultFilter("   aaa", "aaa")).toBe(true);
      expect(defaultFilter("aaa   ", "aaa")).toBe(true);
      expect(defaultFilter("   aaa   ", "aaa")).toBe(true);
    });

    it("negatively filters text that doesn't lexically match", () => {
      expect(defaultFilter("a", "b")).toBe(false);
    });

    it("works case insensitively", () => {
      expect(defaultFilter("a", "A")).toBe(true);
    });
  });

  it("matches snapshot", () => {
    const comp = mountTextInputAutocomplete();
    expect(comp).toMatchSnapshot();
  });

  it("should open dropdown on focus", () => {
    const comp = mountTextInputAutocomplete();
    expect(comp.find("DropdownList").length).toBe(0);
    comp.find("input").simulate("focus");
    expect(comp.find("DropdownList").length).toBe(1);
  });

  it("should open dropdown on mouse down", () => {
    const comp = mountTextInputAutocomplete();
    expect(comp.find("DropdownList").length).toBe(0);
    comp.find("input").simulate("mousedown");
    expect(comp.find("DropdownList").length).toBe(1);
  });

  it("should close dropdown on mouse down if already open", () => {
    const comp = mountTextInputAutocomplete();
    comp.find("input").simulate("focus");
    expect(comp.find("DropdownList").length).toBe(1);
    comp.find("input").simulate("mousedown");
    expect(comp.find("DropdownList").length).toBe(0);
  });

  it("should filter down on options based on value is entered", () => {
    const comp = mountTextInputAutocomplete({
      value: "bbb",
      suggestions: ["aaa", "aaabbb", "aaabbbccc"],
    });
    comp.find("input").simulate("focus");

    expect(comp.find("ul").children().length).toBe(2);
  });

  it("calls onChange when an option is clicked with the respective value", () => {
    const handleChange = jest.fn();

    const comp = mountTextInputAutocomplete({
      value: "",
      suggestions: ["aaa", "aaabbb", "aaabbbccc"],
      onChange: handleChange,
    });

    comp.find("input").simulate("focus");
    const firstOption = comp.find("DropdownOption").first();

    firstOption.simulate("mousedown");

    expect(handleChange).toBeCalled();
    expect(handleChange.mock.calls[0][0]).toBe("aaa");
  });

  it("wont filter options when suggestionsFilter is null", () => {
    const comp = mountTextInputAutocomplete({
      value: "---",
      suggestions: ["aaa", "aaabbb", "aaabbbccc"],
      suggestionsFilter: null,
    });

    comp.find("input").simulate("focus");
    const options = comp.find("DropdownOption");

    expect(options.length).toBe(3);
  });
});
