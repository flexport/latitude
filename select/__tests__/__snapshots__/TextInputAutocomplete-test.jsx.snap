// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputAutocomplete matches snapshot 1`] = `
<TextInputAutocomplete
  maximumOptions={10}
  onChange={[Function]}
  suggestions={
    Array [
      "aaa",
      "aaabbb",
      "aaabbbccc",
    ]
  }
  suggestionsFilter={[Function]}
  value=""
>
  <PopupWithClickAway
    escToClose={false}
    onClose={[Function]}
    onOpen={[Function]}
  >
    <Manager>
      <OnClickOutside(OnClickOutside)
        eventTypes={
          Array [
            "mousedown",
            "touchstart",
          ]
        }
        excludeScrollbar={false}
        onClickOutside={[Function]}
        outsideClickIgnoreClass="ignore-react-onclickoutside"
        preventDefault={false}
        stopPropagation={false}
      >
        <OnClickOutside
          disableOnClickOutside={[Function]}
          enableOnClickOutside={[Function]}
          eventTypes={
            Array [
              "mousedown",
              "touchstart",
            ]
          }
          onClickOutside={[Function]}
          outsideClickIgnoreClass="ignore-react-onclickoutside"
          preventDefault={false}
          stopPropagation={false}
        >
          <div
            className="container_e296pg"
          >
            <Component>
              <Reference>
                <InnerReference
                  setReferenceNode={[Function]}
                >
                  <div
                    onKeyDown={[Function]}
                    role="presentation"
                  >
                    <TextInput
                      disabled={false}
                      isInvalid={false}
                      isPrefilled={false}
                      maxLength={255}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      placeholder=""
                      readOnly={false}
                      size="m"
                      textAlign="left"
                      type="text"
                      value=""
                    >
                      <div
                        className="wrapper_oxngbz"
                      >
                        <input
                          className="baseNoPadding_656545"
                          disabled={false}
                          maxLength={255}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onMouseDown={[Function]}
                          placeholder=""
                          readOnly={false}
                          style={
                            Object {
                              "paddingLeft": 10,
                              "paddingRight": 10,
                              "textAlign": "left",
                            }
                          }
                          type="text"
                          value=""
                        />
                      </div>
                    </TextInput>
                  </div>
                </InnerReference>
              </Reference>
            </Component>
            <Component
              noPortal={true}
              placement="bottom-start"
              zIndex={1500}
            >
              <div
                className="ignore-react-onclickoutside"
              >
                <Popper
                  placement="bottom-start"
                >
                  <InnerPopper
                    eventsEnabled={true}
                    placement="bottom-start"
                    positionFixed={false}
                    referenceElement={
                      <div
                        role="presentation"
                      >
                        <div
                          class="wrapper_oxngbz"
                        >
                          <input
                            class="baseNoPadding_656545"
                            maxlength="255"
                            placeholder=""
                            style="text-align: left; padding-left: 10px; padding-right: 10px;"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    }
                  />
                </Popper>
              </div>
            </Component>
          </div>
        </OnClickOutside>
      </OnClickOutside(OnClickOutside)>
    </Manager>
  </PopupWithClickAway>
</TextInputAutocomplete>
`;
